browser/plugins/test/network-breadcrumbs.test.js

```diff
@@ -0,0 +1,171 @@
+// magical jasmine globals
+const { describe, it, expect, jasmine, XMLHttpRequest, XDomainRequest } = global
+
+const plugin = require('../network-breadcrumbs')
+
+const Client = require('../../../base/client')
+const VALID_NOTIFIER = { name: 't', version: '0', url: 'http://notify.endpoint' }
+
+describe('plugin: network breadcrumbs', () => {
+  if ('addEventListener' in XMLHttpRequest.prototype) {
+    it('should leave a breadcrumb when an XMLHTTPRequest resolves', (done) => {
+      setup((client) => {
+        const request = new XMLHttpRequest()
+        request.open('GET', '/')
+
+        request.addEventListener('load', () => {
+          expect(client.breadcrumbs.length).toBe(1)
+          expect(client.breadcrumbs[0]).toEqual(jasmine.objectContaining({
+            type: 'network',
+            name: 'XMLHttpRequest succeeded',
+            metaData: {
+              status: 200,
+              request: 'GET /'
+            }
+          }))
+          done()
+        })
+
+        request.send()
+      })
+    })
+
+    it('should not leave duplicate breadcrumbs if open() is called twice', (done) => {
+      setup((client) => {
+        const request = new XMLHttpRequest()
+        request.open('GET', '/')
+        request.open('GET', '/')
+
+        request.addEventListener('load', () => {
+          expect(client.breadcrumbs.length).toBe(1)
+          done()
+        })
+
+        request.send()
+      })
+    })
+
+    it('should leave a breadcrumb when an XMLHTTPRequest has a failed response', (done) => {
+      setup((client) => {
+        const request = new XMLHttpRequest()
+        request.open('GET', '/this-does-not-exist')
+
+        request.addEventListener('load', () => {
+          expect(client.breadcrumbs.length).toBe(1)
+          expect(client.breadcrumbs[0]).toEqual(jasmine.objectContaining({
+            type: 'network',
+            name: 'XMLHttpRequest failed',
+            metaData: {
+              status: 404,
+              request: 'GET /this-does-not-exist'
+            }
+          }))
+          done()
+        })
+
+        request.send()
+      })
+    })
+
+    // Browsers that support XDomainRequest will throw a real error when trying to do a cross
+    // domain request. We don't need a breadcrumb for this since bugsnag will track that error
+    if (!XDomainRequest) {
+      it('should leave a breadcrumb when an XMLHTTPRequest has a network error', (done) => {
+        setup((client) => {
+          const request = new XMLHttpRequest()
+          request.open('GET', 'https://api.bugsnag.com')
```

Comment:
does this test rely on https://api.bugsnag.com being accessible and responding correctly every time?

Reply:
This test assumes that the request will fail because of a CORS error